Index: ltmain.sh
===================================================================
RCS file: /export/cvs/projects/SimVoleon/SimVoleon/cfg/ltmain.sh,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 ltmain.sh
--- ltmain.sh	29 Jul 2002 08:53:09 -0000	1.1.1.1
+++ ltmain.sh	14 May 2003 09:44:50 -0000
@@ -2894,7 +2894,18 @@
 	if test -n "$export_symbols" && test -n "$archive_expsym_cmds"; then
 	  eval cmds=\"$archive_expsym_cmds\"
 	else
+         save_deplibs="$deplibs"
+         for conv in $convenience; do
+       tmp_deplibs=
+       for test_deplib in $deplibs; do
+         if test "$test_deplib" != "$conv"; then
+           tmp_deplibs="$tmp_deplibs $test_deplib"
+         fi
+       done
+       deplibs="$tmp_deplibs"
+         done
 	  eval cmds=\"$archive_cmds\"
+    deplibs="$save_deplibs"
 	fi
 	save_ifs="$IFS"; IFS='~'
 	for cmd in $cmds; do
